File: terraform/scripts/automation_schedule_seed.py
#!/usr/bin/env python3
"""Compute first Azure Automation weekday schedule occurrences.

Input JSON on stdin:
  {
    "anchor_rfc3339": "...",
    "timezone": "Europe/London",
    "startup_time": "13:30",
    "shutdown_time": "16:15"
  }

Output JSON on stdout:
  {
    "startup_start": "2026-03-19T13:30:00Z",
    "shutdown_start": "2026-03-19T16:15:00Z"
  }
"""

from __future__ import annotations

import json
import sys
from datetime import UTC, datetime, timedelta
from zoneinfo import ZoneInfo


def parse_hhmm(value: str) -> tuple[int, int]:
    hour_str, minute_str = value.split(":", 1)
    return int(hour_str), int(minute_str)


def next_weekday_occurrence(anchor_utc: datetime, tz_name: str, hhmm: str) -> str:
    tz = ZoneInfo(tz_name)
    anchor_local = anchor_utc.astimezone(tz)
    hour, minute = parse_hhmm(hhmm)

    candidate_date = anchor_local.date()
    while True:
        is_weekday = candidate_date.weekday() < 5
        candidate_local = datetime(
            candidate_date.year,
            candidate_date.month,
            candidate_date.day,
            hour,
            minute,
            tzinfo=tz,
        )

        if is_weekday and anchor_local < candidate_local:
            candidate_utc = candidate_local.astimezone(UTC)
            return candidate_utc.strftime("%Y-%m-%dT%H:%M:%SZ")

        candidate_date += timedelta(days=1)


def main() -> int:
    query = json.load(sys.stdin)
    anchor_rfc3339 = query["anchor_rfc3339"].replace("Z", "+00:00")
    anchor_utc = datetime.fromisoformat(anchor_rfc3339).astimezone(UTC)
    timezone = query["timezone"]

    result = {
        "startup_start": next_weekday_occurrence(anchor_utc, timezone, query["startup_time"]),
        "shutdown_start": next_weekday_occurrence(anchor_utc, timezone, query["shutdown_time"]),
    }

    json.dump(result, sys.stdout)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: terraform/scripts/automation_schedule_seed.pl
#!/usr/bin/env perl
use strict;
use warnings;
use JSON::PP qw(decode_json encode_json);
use POSIX qw(tzset);
use Time::Local qw(timegm timelocal);

sub parse_rfc3339_epoch {
    my ($value) = @_;

    die "invalid RFC3339 timestamp: $value"
        unless $value =~ /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(Z|([+-])(\d{2}):?(\d{2}))$/;

    my ($year, $month, $day, $hour, $minute, $second) = ($1, $2, $3, $4, $5, $6);
    my $offset = 0;
    if ($7 ne 'Z') {
        $offset = ($9 * 3600) + ($10 * 60);
        $offset *= -1 if $8 eq '-';
    }

    return timegm($second, $minute, $hour, $day, $month - 1, $year) - $offset;
}

sub utc_rfc3339 {
    my ($epoch) = @_;
    my @t = gmtime($epoch);
    return sprintf(
        '%04d-%02d-%02dT%02d:%02d:%02dZ',
        $t[5] + 1900, $t[4] + 1, $t[3], $t[2], $t[1], $t[0]
    );
}

sub with_timezone {
    my ($tz, $callback) = @_;

    my $old_tz = $ENV{TZ};
    $ENV{TZ} = $tz;
    tzset();

    my $ok = eval {
        $callback->();
        1;
    };
    my $err = $@;

    if (defined $old_tz) {
        $ENV{TZ} = $old_tz;
    }
    else {
        delete $ENV{TZ};
    }
    tzset();

    die $err unless $ok;
}

sub next_weekday_occurrence {
    my ($anchor_epoch, $tz, $hhmm) = @_;

    my ($hour, $minute) = split /:/, $hhmm, 2;
    die "invalid time format: $hhmm" unless defined $hour && defined $minute;
    die "invalid time format: $hhmm" unless $hour =~ /^\d{1,2}$/ && $minute =~ /^\d{1,2}$/;
    $hour = int($hour);
    $minute = int($minute);
    die "invalid time format: $hhmm" unless $hour >= 0 && $hour <= 23 && $minute >= 0 && $minute <= 59;

    my $result;
    with_timezone($tz, sub {
        my @anchor_local = localtime($anchor_epoch);
        my ($day, $month, $year) = ($anchor_local[3], $anchor_local[4], $anchor_local[5] + 1900);

        while (1) {
            my $candidate_epoch = timelocal(0, $minute, $hour, $day, $month, $year);
            my @candidate_local = localtime($candidate_epoch);
            my $wday = $candidate_local[6] == 0 ? 7 : $candidate_local[6];

            if ($wday >= 1 && $wday <= 5 && $anchor_epoch < $candidate_epoch) {
                $result = utc_rfc3339($candidate_epoch);
                last;
            }

            my $next_noon = timelocal(0, 0, 12, $day + 1, $month, $year);
            my @next_local = localtime($next_noon);
            ($day, $month, $year) = ($next_local[3], $next_local[4], $next_local[5] + 1900);
        }
    });

    return $result;
}

my $stdin = do { local $/ = undef; <STDIN> };
my $query = decode_json($stdin // '{}');

my $anchor_rfc3339 = $query->{anchor_rfc3339} // die "missing anchor_rfc3339";
my $anchor_epoch = parse_rfc3339_epoch($anchor_rfc3339);

my $timezone      = $query->{timezone}      // die "missing timezone";
my $startup_time  = $query->{startup_time}  // die "missing startup_time";
my $shutdown_time = $query->{shutdown_time} // die "missing shutdown_time";

my %result = (
    startup_start  => next_weekday_occurrence($anchor_epoch, $timezone, $startup_time),
    shutdown_start => next_weekday_occurrence($anchor_epoch, $timezone, $shutdown_time),
);

print encode_json(\%result);
